consumer/template_group_writer.go

```diff
@@ -0,0 +1,159 @@
+	"context"
+	"fmt"
+	"github.com/goat-project/goat/consumer/wrapper"
+	"os"
+	"path"
+	"text/template"
+)
+
+const (
+	templateFileExtension = "tmpl"
+	filenameFormat        = "%014d"
+)
+
+type vmsTemplateData struct {
+	Vms []interface{}
+}
+
+// TemplateGroupWriter converts each record to template and writes it to file. Multiple records may be written into a single file.
+type TemplateGroupWriter struct {
+	dir          string
+	templatesDir string
+	countPerFile uint64
+	templateName string
+	template     *template.Template
+	records      []interface{}
+}
+
+// NewTemplateGroupWriter creates a new TemplateGroupWriter.
+func NewTemplateGroupWriter(dir, templatesDir, templateName string, countPerFile uint64) TemplateGroupWriter {
+	return TemplateGroupWriter{
+		dir:          dir,
+		templatesDir: templatesDir,
+		countPerFile: countPerFile,
+		templateName: templateName,
+		template:     nil,
+		records:      make([]interface{}, countPerFile),
+	}
+}
+
+func ensureDirectoryExists(path string) error {
+	err := os.MkdirAll(path, os.ModeDir)
+	if err != nil && err != os.ErrExist {
+		return err
+	}
+
+	return nil
+}
+
+func (tgw *TemplateGroupWriter) initTemplates() error {
+	if tgw.template != nil {
+		return nil
+	}
+
+	template, err := template.ParseGlob(path.Join(tgw.templatesDir, fmt.Sprintf("*.%s", templateFileExtension)))
+	tgw.template = template
+	return err
+}
+
+func trySendError(ctx context.Context, res chan<- Result, err error) {
+	for {
+		select {
+		case res <- NewErrorResult(err):
+			// error sent successfully
+			return
+		case <-ctx.Done():
+			// goroutine has been cancelled
+			return
+		}
+	}
+}
+
+func writeFile(data interface{}, recordCount uint64, template *template.Template, filename, templateName string) error {
+	// open the file
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+
+	// write templateData to file
+	err = template.ExecuteTemplate(file, templateName, data)
+	if err != nil {
+		return err
+	}
+
+	// close file
+	return file.Close()
+}
+
+// Consume converts each record to template and writes it to file. Multiple records may be written into a single file.
+func (tgw TemplateGroupWriter) Consume(ctx context.Context, id string, records <-chan wrapper.RecordWrapper) (ResultsChannel, error) {
+	res := make(chan Result)
+
+	if err := ensureDirectoryExists(path.Join(tgw.dir, id)); err != nil {
+		return nil, err
+	}
+
+	if err := tgw.initTemplates(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+
+		defer close(res)
+
+		var countInFile, filenameCounter uint64
+		countInFile, filenameCounter = 0, 0
+		for {
+			select {
+			case templateData, ok := <-records:
+				if !ok {
+					// end of stream
+					if countInFile > 0 {
+						// but we have something to save!
+						templateData := vmsTemplateData{Vms: tgw.records}
+						err := writeFile(templateData, countInFile, tgw.template, path.Join(tgw.dir, path.Join(id, fmt.Sprintf(filenameFormat, filenameCounter))), tgw.templateName)
```

Comment:
Can't this be a method? You're doing the same thing on line 140 as well.

Reply:
You're right, I'll turn it into a method of `TemplateGroupWriter`.